File: app/src/main/java/malayalamdictionary/samasya/HistoryEnglishFragment.java
package malayalamdictionary.samasya;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.DialogInterface;
import android.database.Cursor;
import android.database.SQLException;
import android.graphics.Point;
import android.os.AsyncTask;
import android.os.Bundle;
import androidx.fragment.app.Fragment;
import android.util.SparseBooleanArray;
import android.view.ActionMode;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.ExpandableListView;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import malayalamdictionary.samasya.adapter.HistoryAdapter;
import malayalamdictionary.samasya.database.DatabaseHelper;
import malayalamdictionary.samasya.helper.HistoryItems;

public class HistoryEnglishFragment extends Fragment {
    //todo history deletion bug,check favorite deletion also

    HistoryAdapter listAdapter;
    ExpandableListView expListView;
    List<HistoryItems> listDataHeader;
    HashMap<String, List<String>> listDataChild;
    List<HistoryItems> historyItems;
    ArrayList<String> meaningArray;
    boolean actionModeEnabled;
    boolean selectAll=true;
    MenuItem menuItem;
    String copyText;
    ProgressDialog progress;
    HistoryTask historyTask;
    public boolean expanded=false;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        View rowView;
        rowView = inflater.inflate(R.layout.fragment_history, container, false);


        expListView = (ExpandableListView)rowView.findViewById(R.id.lvExp_history);
        listDataHeader = new ArrayList<HistoryItems>();
        listDataChild = new HashMap<String, List<String>>();


        Display display = getActivity().getWindowManager().getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);
        int width = size.x;
        expListView.setIndicatorBounds(width-GetDipsFromPixel(35), width-GetDipsFromPixel(5));

        progress = new ProgressDialog(getActivity());
        historyTask=new HistoryTask();
        historyTask.execute();


        expListView.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE_MODAL);

        expListView.setOnGroupClickListener(new ExpandableListView.OnGroupClickListener() {
            @Override
            public boolean onGroupClick(ExpandableListView expandableListView, View view, int i, long l) {

                if(!expandableListView.isGroupExpanded(i))
                {
                    expanded=true;
                }
                else {
                    expanded=false;
                }
                if(actionModeEnabled)
                {
                    expandableListView.setItemChecked(i, !expandableListView.isItemChecked(i));

                }

                return actionModeEnabled;
            }
        });



        expListView.setMultiChoiceModeListener(new AbsListView.MultiChoiceModeListener() {
            @Override
            public void onItemCheckedStateChanged(ActionMode mode, int position, long id, boolean checked) {
                final int checkedCount = expListView.getCheckedItemCount();
                mode.setTitle(String.valueOf(checkedCount));
                listAdapter.toggleSelection(position);
                final SparseBooleanArray selected = listAdapter.getSelectedIds();

                if (checkedCount > 1) {
                    menuItem.setVisible(false);
                } else {

                    if (selected.valueAt(0)) {
                        copyText = String.valueOf(listAdapter.getGroupItem(selected.keyAt(0)));

                    }

                    menuItem.setVisible(true);

                }

                if (checkedCount == listAdapter.getGroupCount()) {
                    selectAll = false;

                } else {
                    selectAll = true;

                }
            }

            @Override
            public boolean onCreateActionMode(ActionMode mode, Menu menu) {
                MenuInflater menuInflater = getActivity().getMenuInflater();
                menuInflater.inflate(R.menu.favourite_menu, menu);
//                toolbar.setVisibility(View.GONE);
                menuItem = menu.getItem(0);
                actionModeEnabled = true;

                return true;
            }

            @Override
            public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
                return false;
            }

            @Override
            public boolean onActionItemClicked(ActionMode mode, MenuItem item) {


                switch (item.getItemId()) {

                    case R.id.copy:
                        if (!expanded) {
                            String label = "copy";
                            ClipboardManager clipboard = (ClipboardManager) getActivity().getSystemService(Context.CLIPBOARD_SERVICE);
                            ClipData clip = ClipData.newPlainText(label, copyText);
                            clipboard.setPrimaryClip(clip);
                            mode.finish();
                            Toast.makeText(getActivity(), copyText + " copied to clipboard", Toast.LENGTH_SHORT).show();
                        }
                        else {
                            Toast.makeText(getActivity(),"Please close expanded words before selection",Toast.LENGTH_SHORT).show();

                        }
                        return true;


                    case R.id.select_all:
                        if (selectAll) {
                            for (int i = 0; i < expListView.getAdapter().getCount(); i++) {
                                expListView.setItemChecked(i, true);
                                listAdapter.selectView(i, true);
                            }
                            selectAll = false;

                        } else {
                            for (int i = 0; i < expListView.getAdapter().getCount(); i++) {

                                expListView.setItemChecked(i, false);
                                listAdapter.selectView(i, false);
                            }
                            selectAll = true;
                        }

                        return true;

                    case R.id.delete:

                        final SparseBooleanArray selected = listAdapter.getSelectedIds();
                        final AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(getActivity());
                        alertDialogBuilder.setMessage("Are you sure,You wanted to delete this item");
                        alertDialogBuilder.setPositiveButton("yes", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface arg0, int arg1) {


//                        // Calls getSelectedIds method from ListViewAdapter Class

//                        // Captures all selected ids with a loop
                                for (int i = (selected.size() - 1); i >= 0; i--) {

                                    DatabaseHelper databaseHelper = new DatabaseHelper(getActivity());
                                    if (selected.valueAt(i)) {

                                        if (!expanded) {

                                            HistoryItems historyItems = (HistoryItems) listAdapter.getGroup(selected.keyAt(i));


                                            listAdapter.remove(historyItems);

                                            try {
                                                Cursor cursor = databaseHelper.getWritableDatabase().rawQuery("delete From samasya_eng_history where ENG like '" + historyItems.getName() + "'", null);
                                                cursor.moveToFirst();
                                                while (!cursor.isAfterLast()) {
                                                   cursor.moveToNext();
                                                }
                                                cursor.close();
                                            } catch (SQLException sqle) {
                                                throw sqle;
                                            }
                                        }
                                    }
                                    else {
                                        Toast.makeText(getActivity(),"Please close expanded words before selection",Toast.LENGTH_SHORT).show();

                                    }
                                }
                            }
                        });

                        alertDialogBuilder.setNegativeButton("No", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                dialog.dismiss();
                            }
                        });

                        AlertDialog alertDialog = alertDialogBuilder.create();
                        alertDialog.show();

                        // Close CAB
                        mode.finish();
                        return true;
                    default:
                        return false;
                }
            }


            @Override
            public void onDestroyActionMode(ActionMode mode) {
//                toolbar.setVisibility(View.VISIBLE);
                listAdapter.removeSelection();
                actionModeEnabled = false;

            }
        });


        return rowView;
    }



    private void prepareListData() {

        getDataFromDb();


        // Adding child

        for (int i=0;i<historyItems.size();i++){
            listDataHeader.add(historyItems.get(i));
            List<String> meaning;
            meaning= getMeaningFromDb(listDataHeader.get(i).getName());
            listDataChild.put(listDataHeader.get(i).getName(), meaning);
        }


    }


    public void getDataFromDb(){

        DatabaseHelper myDbHelper2 = new DatabaseHelper(getActivity());
        try {
            myDbHelper2.open();
            Cursor cursor = myDbHelper2.getReadableDatabase().rawQuery("Select DISTINCT ENG From samasya_eng_history order by ENG asc", null);
            historyItems = new ArrayList<>();
            cursor.moveToFirst();
            while (!cursor.isAfterLast()) {
                historyItems.add(new HistoryItems(cursor.getString(0)));
                cursor.moveToNext();
            }
            cursor.close();
            myDbHelper2.close();
        } catch (SQLException sqle) {
            throw sqle;
        }
    }

    public ArrayList<String> getMeaningFromDb(String word){

        DatabaseHelper myDbHelper2 = new DatabaseHelper(getActivity());
        try {
            myDbHelper2.open();
            Cursor cursor = myDbHelper2.getReadableDatabase().rawQuery("Select * From samasya_eng_mal where ENG like '"+word+"'", null);
            meaningArray = new ArrayList<>();
            cursor.moveToFirst();
            while (!cursor.isAfterLast()) {
                meaningArray.add(cursor.getString(1));
                cursor.moveToNext();
            }
            cursor.close();
            myDbHelper2.close();
        } catch (SQLException sqle) {
            throw sqle;
        }

        return meaningArray;
    }

    public int GetDipsFromPixel(float pixels)
    {
        // Get the screen's density scale
        final float scale = getResources().getDisplayMetrics().density;
        // Convert the dps to pixels, based on density scale
        return (int) (pixels * scale + 0.5f);
    }

    private class HistoryTask extends AsyncTask<String, String, String> {


        @Override
        protected String doInBackground(String... params) {
            prepareListData();

            return null;
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            progress.setMessage("Loading...");
            progress.show();

        }

        @Override
        protected void onPostExecute(String s) {
            super.onPostExecute(s);
            progress.dismiss();
            listAdapter = new HistoryAdapter(getActivity(), listDataHeader, listDataChild,true);
            expListView.setAdapter(listAdapter);

        }
    }


    }

File: app/src/main/java/malayalamdictionary/samasya/HistoryEnglishFragment.kt
package malayalamdictionary.samasya

import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.database.SQLException
import android.graphics.Point
import android.os.AsyncTask
import android.os.Bundle
import android.util.Log
import android.util.SparseBooleanArray
import android.view.*
import android.widget.AbsListView
import android.widget.ExpandableListView
import android.widget.ListView
import android.widget.Toast
import androidx.fragment.app.Fragment
import malayalamdictionary.samasya.adapter.HistoryAdapter
import malayalamdictionary.samasya.database.DatabaseHelper
import malayalamdictionary.samasya.helper.HistoryItems
import java.util.*

class HistoryEnglishFragment : Fragment() {
    //todo history deletion bug,check favorite deletion also
    private lateinit var listAdapter: HistoryAdapter
    private lateinit var expListView: ExpandableListView
    private lateinit var listDataHeader: MutableList<HistoryItems>
    private lateinit var listDataChild: HashMap<String, List<String>>
    private lateinit var historyItems: MutableList<HistoryItems>
    private lateinit var meaningArray: ArrayList<String>
    private var actionModeEnabled: Boolean = false
    private var selectAll = true
    private lateinit var menuItem: MenuItem
    private lateinit var copyText: String
    private lateinit var progress: ProgressDialog
    private lateinit var historyTask: HistoryTask
    private lateinit var expandedItem: SparseBooleanArray
    var expanded = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rowView: View = inflater.inflate(R.layout.fragment_history, container, false)

        expListView = rowView.findViewById(R.id.lvExp_history)
        listDataHeader = ArrayList()
        listDataChild = HashMap()
        expandedItem = SparseBooleanArray()

        val display: Display? = activity?.windowManager?.defaultDisplay
        val size = Point()
        display?.getSize(size)
        val width = size.x
        expListView.setIndicatorBounds(width - (getDipsFromPixel(35F)), width - getDipsFromPixel(5F))

        progress = ProgressDialog(activity)
        historyTask = HistoryTask()
        historyTask.execute()

        expListView.choiceMode = ListView.CHOICE_MODE_MULTIPLE_MODAL


        expListView.setOnGroupClickListener { expandableListView, view, i, l ->
            expanded = !expandableListView.isGroupExpanded(i)
            if (!actionModeEnabled && expanded) {
                expandedItem.put(i, expanded)
            }
            if (!expanded && expandedItem[i]) {
                expandedItem.delete(i)
            }

            if (actionModeEnabled) {
                expandableListView.setItemChecked(i, !expandableListView.isItemChecked(i))
            }
            actionModeEnabled
        }


        expListView.setMultiChoiceModeListener(object : AbsListView.MultiChoiceModeListener {

            override fun onItemCheckedStateChanged(mode: ActionMode, position: Int, id: Long, checked: Boolean) {
                if (expandedItem.size() == 0) {
                    val checkedCount = expListView.checkedItemCount
                    mode.title = checkedCount.toString()
                    listAdapter.toggleSelection(position)
                    val selected = listAdapter.getSelectedIds()

                    if (checkedCount > 1) {
                        menuItem.isVisible = false
                    } else {
                        if (selected.valueAt(0)) {
                            copyText = listAdapter.getGroupItem(selected.keyAt(0)).toString()
                        }
                        menuItem.isVisible = true
                    }
                    selectAll = checkedCount != listAdapter.groupCount
                    Log.d("check_login", "iam in")
                } else {

                    //if any of the item is expanded then refresh the adapter
                    Log.d("check_login", "iam else")

                    Toast.makeText(activity, "Please close expanded words before selection", Toast.LENGTH_SHORT).show()
                    expandedItem.clear()
                    expListView.setAdapter(listAdapter)
                }
            }

            override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
                val menuInflater = activity!!.menuInflater
                menuInflater.inflate(R.menu.favourite_menu, menu)
                menuItem = menu.getItem(0)
                actionModeEnabled = true

                return true
            }

            override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
                return false
            }

            override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {

                when (item.itemId) {

                    R.id.copy -> {
                        val label = "copy"
                        val clipboard = activity!!.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                        val clip = ClipData.newPlainText(label, copyText)
                        clipboard.primaryClip = clip
                        mode.finish()
                        Toast.makeText(activity, "$copyText copied to clipboard", Toast.LENGTH_SHORT).show()

                        return true
                    }


                    R.id.select_all -> {
                        selectAll = if (selectAll) {
                            for (i in 0 until expListView.adapter.count) {
                                expListView.setItemChecked(i, true)
                                listAdapter.selectView(i, true)
                            }
                            false

                        } else {
                            for (i in 0 until expListView.adapter.count) {

                                expListView.setItemChecked(i, false)
                                listAdapter.selectView(i, false)
                            }
                            true
                        }

                        return true
                    }

                    R.id.delete -> {

                        val selected = listAdapter.getSelectedIds()
                        val alertDialogBuilder = AlertDialog.Builder(activity)
                        alertDialogBuilder.setMessage("Are you sure,You wanted to delete this item")
                        alertDialogBuilder.setPositiveButton("yes") { arg0, arg1 ->
                            // Calls getSelectedIds method from ListViewAdapter Class

                            // Captures all selected ids with a loop
                            for (i in selected.size() - 1 downTo 0) {

                                val databaseHelper = DatabaseHelper(requireContext())
                                if (selected.valueAt(i)) {

                                    val historyItems = listAdapter.getGroup(selected.keyAt(i)) as HistoryItems

                                    listAdapter.remove(historyItems)

                                    try {
                                        val cursor = databaseHelper.writableDatabase.rawQuery("delete From samasya_eng_history where ENG like '" + historyItems.name + "'", null)
                                        cursor.moveToFirst()
                                        while (!cursor.isAfterLast) {
                                            cursor.moveToNext()
                                        }
                                        cursor.close()
                                    } catch (sqle: SQLException) {
                                        throw sqle
                                    }

                                }

                            }
                        }

                        alertDialogBuilder.setNegativeButton("No") { dialog, which -> dialog.dismiss() }

                        val alertDialog = alertDialogBuilder.create()
                        alertDialog.show()

                        // Close CAB
                        mode.finish()
                        return true
                    }
                    else -> return false
                }
            }


            override fun onDestroyActionMode(mode: ActionMode) {
                listAdapter.removeSelection()
                actionModeEnabled = false

            }
        })


        return rowView
    }

    private fun getDipsFromPixel(pixels: Float): Int {
        // Get the screen's density scale
        val scale = resources.displayMetrics.density
        // Convert the dps to pixels, based on density scale
        return (pixels * scale + 0.5f).toInt()
    }


    private fun getDataFromDb() {

        val myDbHelper2 = DatabaseHelper(requireContext())
        try {
            myDbHelper2.open()
            val cursor = myDbHelper2.readableDatabase.rawQuery("Select DISTINCT ENG From samasya_eng_history order by ENG asc", null)
            historyItems = ArrayList()
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                historyItems.add(HistoryItems(cursor.getString(0)))
                cursor.moveToNext()
            }
            cursor.close()
            myDbHelper2.close()
        } catch (sqle: SQLException) {
            throw sqle
        }

    }

    private fun getMeaningFromDb(word: String?): ArrayList<String> {

        val myDbHelper2 = DatabaseHelper(requireContext())
        try {
            myDbHelper2.open()
            val cursor = myDbHelper2.readableDatabase.rawQuery("Select * From samasya_eng_mal where ENG like '$word'", null)
            meaningArray = ArrayList()
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                meaningArray.add(cursor.getString(1))
                cursor.moveToNext()
            }
            cursor.close()
            myDbHelper2.close()
        } catch (sqle: SQLException) {
            throw sqle
        }

        return meaningArray
    }

    private fun prepareListData() {

        getDataFromDb()
        // Adding child
        for (i in historyItems.indices) {
            listDataHeader.add(historyItems[i])
            val meaning: List<String>
            meaning = getMeaningFromDb(listDataHeader[i].name)
            listDataChild[listDataHeader[i].name.toString()] = meaning
        }
    }

    private inner class HistoryTask : AsyncTask<String, String, String>() {

        override fun doInBackground(vararg params: String): String? {
            prepareListData()
            return null
        }

        override fun onPreExecute() {
            super.onPreExecute()
            progress.setMessage("Loading...")
            progress.show()

        }

        override fun onPostExecute(s: String?) {
            super.onPostExecute(s)
            progress.dismiss()
            listAdapter = HistoryAdapter(requireContext(), listDataHeader, listDataChild, true)
            expListView.setAdapter(listAdapter)

        }
    }

}